### PR TITLE
Upgrade TSLint to support TS 3.1

### PR DESCRIPTION
**Pull request by bowenni**
#### PR checklist

- [ ] Addresses an existing issue: #0000
- [x] New feature, bugfix, or enhancement
  - [ ] Includes tests
- [ ] Documentation update

#### Overview of change:
Upgrade TSLint to support TS 3.1.6 and also upgrade tsutils to 3.3.1.
Fixes https://github.com/palantir/tslint/pull/4102

#### Is there anything you'd like reviewers to focus on?
TS 2.1 tests are broken by this change:
tsutils v3.0.0+ now requires node >= 6 ([link](https://github.com/ajafff/tsutils/blob/v3.0.0/package.json#L56)). Therefore it's incompatible with node version specified in the build config of TS 2.1 tests ([link](https://github.com/palantir/tslint/blob/5.11.0/.circleci/config.yml#L46)). I have tried to pin TS 2.1 tests to use an old version of tsutils (v2.9.0 is the last version that is compatible with node 4) but due to some breaking changes in tsutils APIs ([link](https://github.com/ajafff/tsutils/releases/tag/v3.0.0)) it's hard to make some TSLint rules work with both the old and the new APIs. Besides, I think supporting multiple versions of the same utility library is not a great idea. Is it possible to relax the node version requirement for TS 2.1 tests?

### CODE REVIEW

**Comment by bowenni** [comment]
Ops I didn't run the linter.
Closing this for now and will re-open when I'm done.

File: src/language/utils.ts [hunk 1]
```diff
@@ -74,7 +74,7 @@ export function isBlockScopedVariable(
     node: ts.VariableDeclaration | ts.VariableStatement,
 ): boolean {
     if (node.kind === ts.SyntaxKind.VariableDeclaration) {
-        const parent = node.parent!;
```
Comment (bowenni) [inline src/language/utils.ts]: A lot of the changes are due to `node.parent` becoming required now. Therefore the non-null assertions are unnecessary.
https://github.com/Microsoft/TypeScript/blob/v3.1.6/lib/typescript.d.ts#L480

File: src/rules/cyclomaticComplexityRule.ts [hunk 2]
```diff
@@ -90,7 +90,7 @@ function walk(ctx: Lint.WalkContext<{ threshold: number }>): void {
     let complexity = 0;
 
     return ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
-        if (isFunctionScopeBoundary(node)) {
+        if (isFunctionScopeBoundary(node) !== ScopeBoundary.None) {
```
Comment (bowenni) [inline src/rules/cyclomaticComplexityRule.ts]: This is due to a breaking change in tsutils API.
https://github.com/ajafff/tsutils/releases/tag/v3.0.0

File: src/language/utils.ts [hunk 3]
```diff
@@ -177,7 +177,10 @@ export function isCombinedNodeFlagSet(node: ts.Node, flagToCheck: ts.NodeFlags):
  *
  * @deprecated no longer used
  */
-export function isCombinedModifierFlagSet(node: ts.Node, flagToCheck: ts.ModifierFlags): boolean {
+export function isCombinedModifierFlagSet(
+    node: ts.Declaration,
```
Comment (bowenni) [inline src/language/utils.ts]: `ts.getCombinedMofidierFlags` takes a `Declaration` now. https://github.com/Microsoft/TypeScript/blob/v3.1.6/lib/typescript.d.ts#L3162

File: src/rules/noShadowedVariableRule.ts [hunk 4]
```diff
@@ -226,10 +226,13 @@ class NoShadowedVariableWalker extends Lint.AbstractWalker<Options> {
...
-            if (boundary === ScopeBoundary.Block) {
-                this.scope = new Scope(parentScope.functionScope);
-            } else if (boundary === ScopeBoundary.Function) {
-                this.scope = new Scope();
+            switch (boundary) {
+                case ScopeBoundary.Block:
+                    this.scope = new Scope(parentScope.functionScope);
+                    break;
+                case ScopeBoundary.Function:
+                case ScopeBoundary.Type:
```
Comment (bowenni) [inline src/rules/noShadowedVariableRule.ts]: tsutils added a new enum member `ScopeBoundary.Type`
https://github.com/ajafff/tsutils/releases/tag/v3.0.0

File: src/rules/noStringThrowRule.ts [hunk 5]
```diff
@@ -71,7 +71,7 @@ function walk(ctx: Lint.WalkContext<void>): void {
     return ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
         if (isThrowStatement(node)) {
             const { expression } = node;
-            if (isString(expression)) {
+            if (expression !== undefined && isString(expression)) {
```
Comment (bowenni) [inline src/rules/noStringThrowRule.ts]: `ThrowStatement` is possibly `undefined` now.
https://github.com/Microsoft/TypeScript/blob/v3.1.6/lib/typescript.d.ts#L1312

**Comment by bowenni** [comment]
@JoshuaKGoldberg Can you please take a look?

**Comment by JoshuaKGoldberg** [comment]
Hi @bowenni, thanks for sending this in!

Most package updates are fine, but changing how TSLint uses tsutils is a breaking change that should be discussed a bit more. A couple examples of it coming into play: https://github.com/Microsoft/tslint-microsoft-contrib/issues/470 https://github.com/Microsoft/tslint-microsoft-contrib/issues/373. How would this change affect `tslint-microsoft-contrib`, for example? It'd be good if we could get consensus from the @palantirtech maintainers of this repo on what the upgrade strategy for dependency versions here is.

Increasing the node version to 6 seems fine, as [4 is no longer supported and 6 is barely in LTS](https://github.com/nodejs/Release).

**Comment by bowenni** [comment]
Thanks @JoshuaKGoldberg . Understood.

I'd love to hear about the deps management strategy from @palantirtech too.

**Comment by bowenni** [comment]
Friendly ping @johnwiseheart

**Comment by NN---** [comment]
I suggest to update this PR for 3.2 support.

**Comment by bowenni** [comment]
The TypeScript version upgrade is an incremental process. Upgrading to 3.1 first will make 3.2 to land more easily.
I do plan to look at the 3.2 upgrade too.

**Comment by malthoff** [comment]
Any progress here?

**Comment by ericanderson** [comment]
Lets not introduce a breaking change on tsutils in case someone was transitively depending on it for their custom rules.

**Comment by bowenni** [comment]
Done. PTAL.

**Comment by bowenni** [comment]
Some mysterious formatting changes are probably due to prettier, I think.

File: yarn.lock [hunk 6]
```diff
@@ -1992,33 +1987,27 @@ tslint-config-prettier@^1.13.0:
...
-  dependencies:
-    tslib "^1.7.1"
+    tslib "^1.8.0"
+    tsutils "^2.27.2"
 
 tsutils@^2.27.2:
-  version "2.27.2"
-  resolved "https://registry.yarnpkg.com/tsutils/-/tsutils-2.27.2.tgz#60ba88a23d6f785ec4b89c6e8179cac9b431f1c7"
-  integrity sha512-qf6rmT84TFMuxAKez2pIfR8UCai49iQsfB7YWVjV1bKpy/d0PWT5rEOSM6La9PiHZ0k1RRZQiwVdVJfQ3BPHgg==
+  version "2.29.0"
```
[the remark this answers is not in the document]
Comment (bowenni) [reply yarn.lock]: Thanks!

**Comment by bowenni** [comment]
Thanks for reviewing the PR!

**Pull request merged**